aten/src/ATen/native/sparse/cuda/SparseBlasImpl.cpp

```diff
@@ -480,6 +481,22 @@ void block_sparse_mm(
       mat1.values().is_contiguous() ||
       mat1.values().transpose(-2, -1).is_contiguous());
 
+  // NOTE: the code below allows arbitrary block sizes
+  // and might be potentially faster than cuSPARSE implementation
+  // especially for not very sparse inputs.
+  if (mat1.scalar_type() == ScalarType::Half || mat1.scalar_type() == ScalarType::BFloat16) {
```

Comment:
This should probably redirect in some more cases: 

From the CuSpares docs (bsrmm):
>The function has the following limitations:
> - Only CUSPARSE_MATRIX_TYPE_GENERAL matrix type is supported
> - Only blockDim > 1 is supported
> - if blockDim ≤ 4, then max(mb)/max(n) = 524,272
> - if 4 < blockDim ≤ 8, then max(mb) = 524,272, max(n) = 262,136
> - if blockDim > 8, then m < 65,535 and max(n) = 262,136

Implicit in this is that b`locksize[0] == blocksize[1] == blockDim`

So if the data type is not supported is already covered, however we should also dispatch to the new implementation if the sparse argument does not have square blocks. 

Additional cases are up for debate:

The size limitations come from the use of 32bit indices. We do not check the sizes, but always convert to 32 bit indices and let CuSparse exit with an error if the size is too large. The limitation is such that if say calling linear with sparse weights and 3d input which is (512, 512, x) input is unfolded to have `n` == 262,144. So with an 8x8 block this call causes an error from cusparse. 

There is also the case of 1x1 block size which is square, but cusparse does not support. Although one could argue that 1x1 blocksize is a misuse of the BSR layout and we should error and the user should use a CSR layout for that case.

Reply:
I think that having 1x1 blocks is totally legit, and what works for csr/csr should work for bsr/bsc with 1x1 blocks.

Reply:
@nikitaved - For optimized performance we'll likely eventually want to vary the block size by operator and overall sparsity pattern. But one step at a time.